README.md
# mesaflash
MesaFlash II
The current version of MesaFlash is this one.

Configuration and diagnostic tool for Mesa Electronics PCI(E)/ETH/EPP/USB/SPI boards

Quickstart:
===========

MesaFlash depends on the libpci-dev, to install it:

  sudo apt install libpci-dev

You may need to install git first:

  sudo apt install git

Clone MesaFlash:

  git clone https://github.com/jethornton/mesaflash.git

From the top level directory, switch to the source directory:

  cd mesaflash

In the source directory to build MesaFlash:

  make

To get command line syntax from a local make:
  
  ./mesaflash --help

To build and install MesaFlash:

  sudo make install

To run an installed MesaFlash:

  mesaflash --help




debian/changelog
mesaflash (4.0.0~alpha0) unstable; urgency=medium

  * Package

 -- Jeff Epler <jepler@unpythonic.net>  Wed, 29 Jan 2020 11:42:12 -0600


mesaflash/common.c
//
//    Copyright (C) 2013-2014 Michael Geszkiewicz
//
//    This program is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation; either version 2 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program; if not, write to the Free Software
//    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
//

#ifdef __linux__
#include <time.h>
#elif _WIN32
#include <windows.h>
#endif
#include <stdio.h>
#include <string.h>
#include "types.h"
#include "common.h"
#include "eeprom.h"

#ifdef _WIN32
void init_io_library() {
    InitializeWinIo();
}

void release_io_library() {
    ShutdownWinIo();
}

u8 inb(u32 addr) {
    u32 val;
    GetPortVal((WORD) addr, &val, 1);
    return val & 0xFF;
}

void outb(u8 data, u32 addr) {
    SetPortVal((WORD) addr, (DWORD) data, 1);
}

u16 inw(u32 addr) {
    u32 val;
    GetPortVal((WORD) addr, &val, 2);
    return val & 0xFFFF;
}

void outw(u16 data, u32 addr) {
    SetPortVal((WORD) addr, (DWORD) data, 2);
}

u32 inl(u32 addr) {
    u32 val;
    GetPortVal((WORD) addr, &val, 4);
    return val;
}

void outl(u32 data, u32 addr) {
    SetPortVal((WORD) addr, (DWORD) data, 4);
}

void *map_memory(u32 base, u32 size, tagPhysStruct_t *phys) {
    void *ptr;

    memset(phys, 0, sizeof(tagPhysStruct_t));
    phys->pvPhysAddress = (DWORD64)(DWORD32) base;
    phys->dwPhysMemSizeInBytes = size;

    ptr = MapPhysToLin(phys);
    return ptr;
}

void *unmap_memory(tagPhysStruct_t *phys) {
    UnmapPhysicalMemory(phys);
}
#endif

void sleep_ns(u64 nanoseconds) {
#ifdef __linux__
    struct timespec tv, tvret;

    tv.tv_sec = 0;
    tv.tv_nsec = nanoseconds;
    nanosleep(&tv, &tvret);
#elif _WIN32
    Sleep(nanoseconds/1000/1000);
#endif
}

void show_formatted_size(u32 size) {
    if (size < 1024) {
        printf(" [size=%u]", size);
    } else if (size < 1024*1024) {
        printf(" [size=%uK]", size/1024);
    } else if (size < 0x80000000) {
        printf(" [size=%uM]", size/(1024*1024));
    } else {
        printf(" [size=%08X]", size);
    }
}

void board_init_struct(board_t *board) {
    memset(board, 0, sizeof(board_t));
    board->llio.board = board;
    board->llio.hm2.llio = &(board->llio);
}

void show_board_info(board_t *board) {
    int i;

    printf("Board info:\n");
    if (board->flash_id > 0) {
        printf("  Flash size: %s (id: 0x%02X)\n", eeprom_get_flash_type(board->flash_id), board->flash_id);
    }
    printf("  Port count: %d\n", board->llio.num_ioport_connectors);
    printf("  Pins per Port: %d\n", board->llio.pins_per_connector);
    printf("  Port names:");
    for (i = 0; i < board->llio.num_ioport_connectors; i++) {
        printf(" %s", board->llio.ioport_connector_name[i]);
    }
    printf("\n");
    printf("  FPGA type: %s\n", board->llio.fpga_part_number);
    printf("  Number of leds: %d\n", board->llio.num_leds);
}


mesaflash/spi_boards.c
//
//    Copyright (C) 2013-2014 Michael Geszkiewicz
//
//    This program is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation; either version 2 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program; if not, write to the Free Software
//    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
//

#include <ctype.h>
#include <fcntl.h>
#include <linux/spi/spidev.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include "types.h"
#include "eeprom.h"
#include "eeprom_local.h"
#include "spi_boards.h"
#include "common.h"

extern board_t boards[MAX_BOARDS];
extern int boards_count;

bool canDo32 = true;

int sd = -1;
struct spi_ioc_transfer settings;

static int spidev_set_lsb_first(int fd, u8 lsb_first) {
    return ioctl(fd, SPI_IOC_WR_LSB_FIRST, &lsb_first);
}

static int spidev_set_mode(int fd, u8 mode) {
    return ioctl(fd, SPI_IOC_WR_MODE, &mode);
}

static int spidev_set_max_speed_hz(int fd, u32 speed_hz) {
    return ioctl(fd, SPI_IOC_WR_MAX_SPEED_HZ, &speed_hz);
}

static int spidev_set_bits_per_word(int fd, u8 bits) {
    return ioctl(fd, SPI_IOC_WR_BITS_PER_WORD, &bits);
}

static u32 read_command(u16 addr, unsigned nelem) {
    bool increment = true;
    return (addr << 16) | SPILBP_CMD_READ | (increment ? SPILBP_ADDR_AUTO_INC : 0) | (nelem << 4);
}

static u32 write_command(u16 addr, unsigned nelem) {
    bool increment = true;
    return (addr << 16) | SPILBP_CMD_WRITE | (increment ? SPILBP_ADDR_AUTO_INC : 0) | (nelem << 4);
}

static int spi_board_open(board_t *board) {
    eeprom_init(&(board->llio));
    board->flash_id = read_flash_id(&(board->llio));
    if (board->fallback_support == 1) {
        eeprom_prepare_boot_block(board->flash_id);
        board->flash_start_address = eeprom_calc_user_space(board->flash_id);
    } else {
        board->flash_start_address = 0;
    }
    return 0;
}

static int spi_board_close(board_t *board) {
    return 0;
}

int spi_boards_init(board_access_t *access) {
    settings.speed_hz = 20 * 1000 * 1000;
    settings.bits_per_word = 32;

    sd = open(access->dev_addr, O_RDWR);
    if(sd == -1) {
        perror("open");
        return -1;
    }
    spidev_set_lsb_first(sd, false);
    spidev_set_mode(sd, 0);
    if (spidev_set_bits_per_word(sd, settings.bits_per_word) < 0)
    {
        fprintf(stderr,"unable to set bpw32, fallback to bpw8\n");
        canDo32 = false;
        settings.bits_per_word = 8;
        spidev_set_bits_per_word(sd, settings.bits_per_word);        
    }
    spidev_set_max_speed_hz(sd, settings.speed_hz);

    return 0;
}

void spi_boards_cleanup(board_access_t *access) {
    if(sd != -1) close(sd);
}

void reorderBuffer(char *pBuf, int numInts)
{
    int lcv;
    for (lcv = 0; lcv < numInts; lcv++)
    {
        pBuf[0] ^= pBuf[3];
        pBuf[3] ^= pBuf[0];
        pBuf[0] ^= pBuf[3];
        pBuf[1] ^= pBuf[2];
        pBuf[2] ^= pBuf[1];
        pBuf[1] ^= pBuf[2];
        pBuf += 4;
    }
}

int spi_read(llio_t *self, u32 addr, void *buffer, int size) {
    if(size % 4 != 0) return -1;
    int numInts = 1+size/4;
    u32 trxbuf[numInts];

    trxbuf[0] = read_command(addr, size/4);
    memset(trxbuf+1, 0, size);
    struct spi_ioc_transfer t;
    t = settings;
    t.tx_buf = t.rx_buf = (uint64_t)(intptr_t)trxbuf;
    t.len = sizeof(trxbuf);
    if (!canDo32)
    {
        reorderBuffer((char *) trxbuf, 1);
    }

    int r = ioctl(sd, SPI_IOC_MESSAGE(1), &t);

    if(r < 0) return r;

    if (!canDo32)
    {
        reorderBuffer((char *) trxbuf, numInts);
    }

    memcpy(buffer, trxbuf+1, size);
    return 0;
}

int spi_write(llio_t *self, u32 addr, void *buffer, int size) {
    if(size % 4 != 0) return -1;
    int numInts = 1+size/4;
    u32 txbuf[numInts];

    txbuf[0] = write_command(addr, size/4);
    memcpy(txbuf + 1, buffer, size);

    struct spi_ioc_transfer t;
    t = settings;
    t.tx_buf = (uint64_t)(intptr_t)txbuf;
    t.len = sizeof(txbuf);
    if (!canDo32)
    {
        reorderBuffer((char *) txbuf, numInts);
    }

    int r = ioctl(sd, SPI_IOC_MESSAGE(1), &t);
    if(r <= 0) return r;
    return 0;
}

static int spi_board_reload(llio_t *self, int fallback_flag) {
    board_t *board = self->board;
    int i;
    u32 boot_addr, cookie;

    spi_read(&(board->llio), HM2_ICAP_REG, &cookie, sizeof(u32));
    if (cookie != HM2_ICAP_COOKIE) {
        printf("ERROR: Active firmware too old to support --reload\n");
        return -1;
    }

    if (fallback_flag == 1) {
        boot_addr = 0x10000;
    } else {
        boot_addr = 0x0;
    }
    boot_addr |= 0x0B000000; // plus read command in high byte

    u32 data[14] = {
      0xFFFF,   // dummy
      0xFFFF,   // dummy
      0xAA99,   // sync
      0x5566,   // sync
      0x3261,   // load low flash start address
      boot_addr & 0xFFFF,   // start addr
      0x3281,   // load high start address + read command
      boot_addr >> 16,   // start addr (plus read command in high byte)
      0x30A1,   // load command register
      0x000E,   // IPROG command
      0x2000,  // NOP
      0x2000,  // NOP
      0x2000,  // NOP
      0x2000  // NOP
    };


    for (i = 0; i < 14; i++) {
        spi_write(&(board->llio), HM2_ICAP_REG, &data[i], sizeof(u32));
        usleep(1000);
    }
    printf("Waiting for FPGA configuration...");
    sleep(2);
    printf("OK\n");
    return 0;
}

void spi_boards_scan(board_access_t *access) {
    u32 buf[4];
    u32 cookie[] = {0x55aacafe, 0x54534f48, 0x32544f4d};
    board_t *board = &boards[boards_count];

    board_init_struct(board);
    if (spi_read(&(board->llio), HM2_COOKIE_REG, &buf, sizeof(buf)) < 0) {
        return;
    }

    if(memcmp(buf, cookie, sizeof(cookie))) {
        fprintf(stderr, "Unexpected cookie at %04x..%04lx:\n%08x %08x %08x\n",
            HM2_COOKIE_REG, HM2_COOKIE_REG + sizeof(buf),
            buf[0], buf[1], buf[2]);
        return;
    }

    char ident[8];
    if(spi_read(&(board->llio), buf[3] + 0xc, &ident, sizeof(ident)) < 0) return;
    
    if(!memcmp(ident, "MESA7I90", 8)) {
        board_t *board = &boards[boards_count];
        board->type = BOARD_SPI;
        strcpy(board->dev_addr, access->dev_addr);
        strncpy(board->llio.board_name, "7I90", 4);
        board->llio.num_ioport_connectors = 3;
        board->llio.pins_per_connector = 24;
        board->llio.ioport_connector_name[0] = "P1";
        board->llio.ioport_connector_name[1] = "P2";
        board->llio.ioport_connector_name[2] = "P3";
        board->llio.num_leds = 2;
        board->llio.verbose = access->verbose;
        board->llio.write = spi_write;
        board->llio.read = spi_read;
        board->llio.write_flash = local_write_flash;
        board->llio.verify_flash = local_verify_flash;
        board->llio.reload = &spi_board_reload;
        board->llio.fpga_part_number = "6slx9tqg144";
        board->open = spi_board_open;
        board->close = spi_board_close;
        board->print_info = spi_print_info;
        board->flash = BOARD_FLASH_HM2;
        board->fallback_support = 1;
        boards_count ++;
    }
        else if(!memcmp(ident, "MESA7C80", 8)) {
        board_t *board = &boards[boards_count];
        board->type = BOARD_SPI;
        strcpy(board->dev_addr, access->dev_addr);
        strncpy(board->llio.board_name, "7C80", 4);
        board->llio.num_ioport_connectors = 2;
        board->llio.pins_per_connector = 27;
        board->llio.ioport_connector_name[0] = "StepGens+Misc";
        board->llio.ioport_connector_name[1] = "Outputs+P1";
        board->llio.num_leds = 4;
        board->llio.verbose = access->verbose;
        board->llio.write = spi_write;
        board->llio.read = spi_read;
        board->llio.write_flash = local_write_flash;
        board->llio.verify_flash = local_verify_flash;
        board->llio.reload = &spi_board_reload;
        board->llio.fpga_part_number = "6slx9tqg144";
        board->open = spi_board_open;
        board->close = spi_board_close;
        board->print_info = spi_print_info;
        board->flash = BOARD_FLASH_HM2;
        board->fallback_support = 1;
        boards_count ++; 
     }   
        else if(!memcmp(ident, "MESA7C81", 8)) {
        board_t *board = &boards[boards_count];
        board->type = BOARD_SPI;
        strcpy(board->dev_addr, access->dev_addr);
        strncpy(board->llio.board_name, "7C81", 4);
        board->llio.num_ioport_connectors = 3;
        board->llio.pins_per_connector = 19;
        board->llio.ioport_connector_name[0] = "P1+Serial";
        board->llio.ioport_connector_name[1] = "P2+Serial";
        board->llio.ioport_connector_name[2] = "P7+Serial";
        board->llio.num_leds = 4;
        board->llio.verbose = access->verbose;
        board->llio.write = spi_write;
        board->llio.read = spi_read;
        board->llio.write_flash = local_write_flash;
        board->llio.verify_flash = local_verify_flash;
        board->llio.reload = &spi_board_reload;
        board->llio.fpga_part_number = "6slx9tqg144";
        board->open = spi_board_open;
        board->close = spi_board_close;
        board->print_info = spi_print_info;
        board->flash = BOARD_FLASH_HM2;
        board->fallback_support = 1;
        boards_count ++; 
     }   
        else {
        int i=0;
        for(i=0; i<sizeof(ident); i++)
            if(!isprint(ident[i])) ident[i] = '?';

        fprintf(stderr, "Unknown board: %.8s\n", ident);
    }
}

void spi_print_info(board_t *board) {
}


mesaflash/mesaflash.c
//
//    Copyright (C) 2013-2014 Michael Geszkiewicz
//
//    This program is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation; either version 2 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program; if not, write to the Free Software
//    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
//

#include <string.h>
#include <stdio.h>
#include <getopt.h>
#include <ctype.h>
#include <stdlib.h>
#include <errno.h>
#include "anyio.h"
#ifdef __linux__
#include <pci/pci.h>
#elif _WIN32
#include "libpci/pci.h"
#endif

static int device_flag;
static int addr_flag;
static int addr_hi_flag;
static int write_flag;
static int fix_boot_flag;
static int verify_flag;
static int auto_verify_flag = 1;
static int fallback_flag;
static int recover_flag;
static int program_flag;
static int readhmid_flag;
static int print_pd_flag;
static int reload_flag;
static int reset_flag;
static int sserial_flag;
static int epp_flag;
static int usb_flag;
static int spi_flag;
static int serial_flag;
static int rpo_flag;
static int wpo_flag;
static u16 rpo_addr;
static u16 wpo_addr;
static u32 wpo_data;
static int set_flag;
static int xml_flag;
static char *lbp16_set_ip_addr;
static int info_flag;
static int verbose_flag;
static char bitfile_name[255];
static board_access_t access;

static struct option long_options[] = {
    {"device", required_argument, 0, 'd'},
    {"addr", required_argument, 0, 'a'},
    {"addr_hi", required_argument, 0, 'b'},
    {"write", required_argument, 0, 'w'},
    {"no-auto-verify", no_argument, &auto_verify_flag, 0},
    {"fix-boot-block", no_argument, &fix_boot_flag, 1},
    {"verify", required_argument, 0, 'v'},
    {"fallback", no_argument, &fallback_flag, 1},
    {"recover", no_argument, &recover_flag, 1},
    {"program", required_argument, 0, 'p'},
    {"readhmid", no_argument, &readhmid_flag, 1},
    {"print-pd", no_argument, &print_pd_flag, 1},
    {"reload", no_argument, &reload_flag, 1},
    {"reset", no_argument, &reset_flag, 1},
    {"sserial", no_argument, &sserial_flag, 1},
    {"epp", no_argument, &epp_flag, 1},
    {"usb", no_argument, &usb_flag, 1},
    {"spi", no_argument, &spi_flag, 1},
    {"serial", no_argument, &serial_flag, 1},
    {"rpo", required_argument, 0, 'r'},
    {"wpo", required_argument, 0, 'o'},
    {"set", required_argument, 0, 's'},
    {"xml", no_argument, &xml_flag, 1},
    {"info", required_argument, 0, 'i'},
    {"help", no_argument, 0, 'h'},
    {"verbose", no_argument, &verbose_flag, 1},
    {0, 0, 0, 0}
};

void print_short_usage() {
    printf("\nMesaflash version 3.4.0~pre (built on %s %s with libpci %s)\n", __DATE__, __TIME__, PCILIB_VERSION);
    printf("Configuration and diagnostic tool for Mesa Electronics PCI(E)/ETH/EPP/USB boards\n");
    printf("(C) 2013-2015 Michael Geszkiewicz (contact: micges@wp.pl)\n");
    printf("(C) Mesa Electronics\n");
    printf("Try 'mesaflash --help' for more information\n");
}

void print_usage() {
    printf("Syntax:\n");
    printf("  mesaflash --device device_name [options]\n");
    printf("  mesaflash --device device_name [options] --write filename\n");
    printf("  mesaflash --device device_name [options] --verify filename\n");
    printf("  mesaflash --device device_name [options] --program filename\n");
    printf("  mesaflash --device device_name [options] --readhmid\n");
    printf("  mesaflash --device device_name [options] --reload | --reset\n");
    printf("  mesaflash --device device_name [options] --sserial\n");
    printf("  mesaflash --device device_name [options] --rpo address\n");
    printf("  mesaflash --device device_name [options] --wpo address=value\n");
    printf("  mesaflash --device device_name [options] --set ip=n.n.n.n\n");
    printf("  mesaflash --info file_name\n");
    printf("  mesaflash --help\n");
    printf("\n");
    printf("Options:\n");
    printf("  --device <name>   Select active device name. If no command is given it\n");
    printf("                    will detect board with given name and print info\n");
    printf("                    about it.\n");
    printf("  --addr <dev>      Select <dev> for looking for <name> (IP address for\n");
    printf("                    Ethernet boards, serial device name for USB boards\n");
    printf("                    and serial boards, SPI device name for SPI boards)\n");
    printf("  --addr_hi         Set the high register address for the EPP interface.\n");
    printf("  --epp             Use EPP interface to connect to board, only for boards\n");
    printf("                    with multiple interfaces (7c80, 7c81, 7i43, 7i90, 7i64).\n");
    printf("  --usb             Use USB interface to connect to board, only for boards\n");
    printf("                    with multiple interfaces (7c80, 7c81, 7i43, 7i90, 7i64).\n");
    printf("  --spi             Use SPI interface to connect to board, only for boards\n");
    printf("                    with multiple interfaces (7c80, 7c81, 7i43, 7i90, 7i64).\n");
    printf("  --serial          Use serial interface to connect to board, only for\n");
    printf("                    boards with multiple interfaces (7i43, 7i90, 7i64).\n");
    printf("  --fallback        Use the fallback area of the EEPROM while executing\n");
    printf("                    commands.\n");
    printf("  --recover         Access board using PCI bridge GPIO (currently\n");
    printf("                    only 6I24/6I25).\n");
    printf("  --xml             Format output from 'readhmid' command into XML.\n");
    printf("  --verbose         Print detailed information while running commands.\n");
    printf("\n");
    printf("Commands:\n");
    printf("  --write           Writes a standard bitfile 'filename' configuration to\n");
    printf("                    the userarea of the EEPROM (IMPORTANT! 'filename' must\n");
    printf("                    be VALID FPGA configuration file).\n");
    printf("  --fix-boot-block  If a write operation does not detect a valid boot\n");
    printf("                    block, write one.\n");
    printf("  --no-auto-verify  Don't automatically verify after writing.\n");
    printf("  --verify          Verifies the EEPROM configuration against the\n");
    printf("                    bitfile 'filename'.\n");
    printf("  --program         Writes a standard bitfile 'filename' configuration to\n");
    printf("                    the FPGA (IMPORTANT! 'filename' must be VALID FPGA\n");
    printf("                    configuration file).\n");
    printf("  --readhmid        Print hostmot2 configuration in PIN file format.\n");
    printf("  --print-pd        Print hostmot2 Pin Descriptors.\n");
    printf("  --reload          Do full FPGA reload from flash (only Ethernet, SPI and\n");
    printf("                    PCI boards).\n");
    printf("  --reset           Do full firmware reset (only Ethernet and serial boards).\n");
    printf("  --sserial         Print full information about all sserial remote boards.\n");
    printf("  --rpo             Read hostmot2 variable directly at 'address'.\n");
    printf("  --wpo             Write hostmot2 variable directly at 'address'\n");
    printf("                    with 'value'.\n");
    printf("  --set             Set board IP address in eeprom to n.n.n.n (only\n");
    printf("                    Ethernet boards).\n");
    printf("  --info            Print info about configuration in 'file_name'.\n");
    printf("  --help            Print this help message.\n");
}

int process_cmd_line(int argc, char *argv[]) {
    int c;

    while (1) {
        /* getopt_long stores the option index here. */
        int option_index = 0;

        c = getopt_long(argc, argv, "", long_options, &option_index);

        /* Detect the end of the options. */
        if (c == -1) break;
        switch (c) {
            case 0: {
                /* If this option set a flag, do nothing else now. */
                if (long_options[option_index].flag != 0) break;
                printf ("option %s", long_options[option_index].name);
                if (optarg)
                    printf (" with arg %s", optarg);
                printf ("\n");
            }
            break;

            case 'd': {
                int i;

                if (device_flag > 0) {
                    printf("Error: multiple --device options\n");
                    exit(-1);
                }
                access.device_name = optarg;
                for (i = 0; optarg[i] != '\0'; i++)
                    access.device_name[i] = toupper(access.device_name[i]);

                device_flag++;
            }
            break;

            case 'a': {
                if (addr_flag > 0) {
                    printf("Error: multiple --addr options\n");
                    exit(-1);
                }
                access.dev_addr = optarg;
                addr_flag++;
            }
            break;

	        case 'b': {
                if (addr_hi_flag > 0) {
                    printf("Error: multiple --addr_hi options\n");
                    exit(-1);
                }
                access.dev_hi_addr = optarg;
                addr_hi_flag++;
            }
	        break;

            case 'w': {
                if (write_flag > 0) {
                    printf("Error: multiple --write options\n");
                    exit(-1);
                }
                strncpy(bitfile_name, optarg, 255);
                write_flag++;
            }
            break;

            case 'p': {
                if (program_flag > 0) {
                    printf("Error: multiple --program options\n");
                    exit(-1);
                }
                strncpy(bitfile_name, optarg, 255);
                program_flag++;
            }
            break;

            case 'r': {
                if (rpo_flag > 0) {
                    printf("Error: multiple --rpo options\n");
                    exit(-1);
                }
                if (strncmp(optarg, "0x", 2) == 0) {
                    optarg[0] = '0';
                    optarg[1] = '0';
                    rpo_addr = strtol(optarg, NULL, 16);
                } else {
                    rpo_addr = strtol(optarg, NULL, 10);
                }
                rpo_flag++;
            }
            break;

            case 'o': {
                char *pch;

                if (wpo_flag > 0) {
                    printf("Error: multiple --wpo options\n");
                    exit(-1);
                }
                pch = strtok(optarg, "=");
                if (strncmp(pch, "0x", 2) == 0) {
                    pch[0] = '0';
                    pch[1] = '0';
                    wpo_addr = strtol(pch, NULL, 16);
                } else {
                    wpo_addr = strtol(pch, NULL, 10);
                }
                pch = strtok(NULL, "=");
                if (strncmp(pch, "0x", 2) == 0) {
                    pch[0] = '0';
                    pch[1] = '0';
                    wpo_data = strtol(pch, NULL, 16);
                } else {
                    wpo_data = strtol(pch, NULL, 10);
                }
                wpo_flag++;
            }
            break;

            case 's': {
                if (set_flag > 0) {
                    printf("Error: multiple --set options\n");
                    exit(-1);
                }
                if (strncmp(optarg, "ip=", 3) == 0) {
                    char *pch;

                    pch = strtok(optarg, "=");
                    pch = strtok(NULL, "=");
                    lbp16_set_ip_addr = pch;
                } else {
                    printf("Error: Unknown set command syntax, see --help for examples\n");
                    exit(-1);
                }
                set_flag++;
            }
            break;

            case 'v': {
                if (verify_flag > 0) {
                    printf("Error: multiple --verify options\n");
                    exit(-1);
                }
                strncpy(bitfile_name, optarg, 255);
                verify_flag++;
            }
            break;

            case 'i': {
                if (info_flag > 0) {
                    printf("Error: multiple --info options\n");
                    exit(-1);
                }
                strncpy(bitfile_name, optarg, 255);
                info_flag++;
            }
            break;

            case 'h': {
                print_usage();
                exit(0);
            }
            break;

            case '?':
                /* getopt_long already printed an error message. */
                return -1;
                break;

            default:
                abort();
        }
    }

    access.type = BOARD_ANY;
    if ((epp_flag == 1) && (usb_flag == 0) && (spi_flag == 0) && (serial_flag == 0)) {
        access.type = BOARD_EPP;
    } else if ((epp_flag == 0) && (usb_flag == 1) && (spi_flag == 0) && (serial_flag == 0)) {
        access.type = BOARD_USB;
    } else if ((epp_flag == 0) && (usb_flag == 0) && (spi_flag == 1) && (serial_flag == 0)) {
        access.type = BOARD_SPI;
    } else if ((epp_flag == 0) && (usb_flag == 0) && (spi_flag == 0) && (serial_flag == 1)) {
        access.type = BOARD_SER;
    }

    return 0;
}

int main(int argc, char *argv[]) {
    int ret = 0;

    if (argc == 1) {
        print_short_usage();
        return 0;
    }
    if (process_cmd_line(argc, argv) == -1)
        return -1;

    if (info_flag == 1) {
        anyio_bitfile_print_info(bitfile_name, 1);
    } else if (device_flag == 1) {
        board_t *board = NULL;

        access.verbose = verbose_flag;
        access.recover = recover_flag;
        access.address = addr_flag;

        if (anyio_init(&access) != 0)
            exit(1);
        ret = anyio_find_dev(&access);
        if (ret < 0) {
            printf("No %s board found\n", access.device_name);
            return -1;
        }
        board = anyio_get_dev(&access, 1);
        if (board == NULL) {
            printf("No %s board found\n", access.device_name);
            return -1;
        }

        board->open(board);

        if (readhmid_flag == 1) {
            anyio_dev_print_hm2_info(board, xml_flag);
        } else if (print_pd_flag == 1) {
            anyio_dev_print_pin_descriptors(board);
        } else if (sserial_flag == 1) {
            anyio_dev_print_sserial_info(board);
        } else if (rpo_flag == 1) {
            u32 data;

            board->llio.read(&(board->llio), rpo_addr, &data, sizeof(u32));
            printf("%08X\n", data);
        } else if (wpo_flag == 1) {
            board->llio.write(&(board->llio), wpo_addr, &wpo_data, sizeof(u32));
        } else if (set_flag == 1) {
            ret = anyio_dev_set_remote_ip(board, lbp16_set_ip_addr);
        } else if (write_flag == 1) {
            ret = anyio_dev_write_flash(board, bitfile_name, fallback_flag, fix_boot_flag);
            if (ret == 0) {
                if (auto_verify_flag) {
                    ret = anyio_dev_verify_flash(board, bitfile_name, fallback_flag);
                }
            }
            if (ret == 0) {
                if (reload_flag == 1) {
                    ret = anyio_dev_reload(board, fallback_flag);
                    if (ret == -1) {
                        printf("\nYou must power cycle the hardware to load a new firmware.\n");
                    }
                } else if (board->llio.reload) {
                    printf("\nYou must power cycle the hardware or use the --reload command to load a new firmware.\n");
                } else {
                    printf("\nYou must power cycle the hardware\n");
                }
            }
        } else if (verify_flag == 1) {
            ret = anyio_dev_verify_flash(board, bitfile_name, fallback_flag);
        } else if (program_flag == 1) {
            ret = anyio_dev_program_fpga(board, bitfile_name);
        } else if (reload_flag == 1) {
            anyio_dev_reload(board, fallback_flag);
        } else if (reset_flag == 1) {
            anyio_dev_reset(board);
        } else {
           board->print_info(board);
        }
        board->close(board);
        anyio_cleanup(&access);
    } else {
        printf("No action requested. Please specify at least --device or --info.\n");
    }
    return ret;
}
